src/app-layer-detect-proto.c

```diff
@@ -532,7 +532,11 @@ static AppProto AppLayerProtoDetectPPGetProto(Flow *f,
     } else {
         /* first try the destination port */
         pp_port_dp = AppLayerProtoDetectGetProbingParsers(alpd_ctx.ctx_pp, ipproto, dp);
-        alproto_masks = &f->probing_parser_toclient_alproto_masks;
+        if (dir == idir) {
```

Comment:
I think I sort of understand what you're trying to do, but I don't understand why it work work or how to follow the flow of things.

Can you update the commit message to explain how in your test case the masks are wrong (or harmful), and what this patch does to address that?

Reply:
Ok doing

Reply:
Well, I had already done it 
```
    Do not mask protocols on both directions with only first packet
    
    For instance :
    When the first packet is no valid DNS but on port 53 (a junk request)
    second packet (error response from server) does not get checked for DNS
    as first packet bit masked away DNS for both directions
```

So 

>  how in your test case the masks are wrong (or harmful)

It is wrong to mask protocols on _both_ directions with only _first_ packet

> what this patch does to address that?

The patch applies the bit mask only for the right direction